File: 113_Path_Sum_II/113_2.cpp
class Solution {
public:
    vector<vector<int>> pathSum2(TreeNode *root, int sum) {
        if (!root) return vector<vector<int>>();
        vector<vector<int>> res2;
        stack<pair<TreeNode*, vector<int>>> s;
        s.push({root, {root->val}});
        while (!s.empty()) {
            pair<TreeNode*, vector<int>> p = s.top(); s.pop();
            int sumVector = accumulate(p.second.begin(), p.second.end(), 0);
            if (!p.first->left && !p.first->right && sum == sumVector)
                res2.emplace_back(p.second);
            if (p.first->left) {
                vector<int> tmp{p.second};
                tmp.emplace_back(p.first->left->val);
                s.push({p.first->left, tmp});
            }
            if (p.first->right) {
                vector<int> tmp{p.second};
                tmp.emplace_back(p.first->right->val);
                s.push({p.first->right, tmp});
            }
        }
        return res2;
    }

    vector<vector<int>> pathSum3(TreeNode *root, int sum) {
        if (!root) return vector<vector<int>>();
        vector<vector<int>> res3;
        stack<TreeNode*> treeNodeStack;
        stack<vector<int>> valStack;
        treeNodeStack.push(root);
        valStack.push({root->val});
        while (!treeNodeStack.empty()) {
            TreeNode *node = treeNodeStack.top(); treeNodeStack.pop();
            vector<int> vals = valStack.top(); valStack.pop();
            int sumVector = accumulate(vals.begin(), vals.end(), 0);
            if (!node->left && !node->right && sum == sumVector)
                res3.emplace_back(vals);
            if (node->left) {
                vector<int> tmp = vals;
                tmp.emplace_back(node->left->val);
                treeNodeStack.push(node->left);
                valStack.push(tmp);
            }
            if (node->right) {
                vector<int> tmp{vals};
                tmp.emplace_back(node->right->val);
                treeNodeStack.push(node->right);
                valStack.push(tmp);
            }
        }
        return res3;
    }

private:
    vector<vector<int>> res;
};

File: 113_Path_Sum_II/113_2.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum_: int) -> List[List[int]]:
        if not root:
            return []
        res = []
        stack = [([root.val], root)]
        while stack:
            tmp, node = stack.pop()
            if not node.left and not node.right and sum(tmp) == sum_:
                res.append(tmp)
            if node.left:
                stack.append((tmp + [node.left.val], node.left))
            if node.right:
                stack.append((tmp + [node.right.val], node.right))
        return res


if __name__ == "__main__":
    node1 = TreeNode(5)
    node2 = TreeNode(4)
    node3 = TreeNode(8)
    node4 = TreeNode(11)
    node5 = TreeNode(13)
    node6 = TreeNode(4)
    node7 = TreeNode(7)
    node8 = TreeNode(2)
    node9 = TreeNode(5)
    node10 = TreeNode(1)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node3.left = node5
    node3.right = node6
    node4.left = node7
    node4.right = node8
    node6.left = node9
    node6.right = node10
    s = Solution()
    print(s.pathSum(node1, 22))